internal/proxy/channels_mgr.go

```diff
@@ -111,323 +117,175 @@ const (
 		return err
 	}
-	stream.AsProducer(pchans)
-	if mgr.repackFunc != nil {
-		stream.SetRepackFunc(mgr.repackFunc)
-	}
-	runtime.SetFinalizer(stream, func(stream msgstream.MsgStream) {
-		stream.Close()
-	})
-	mgr.addStream(id, stream)
 
-	mgr.updateCollection(collectionID, id)
-	for _, pc := range pchans {
+	mgr.updateCollection(collectionID, channelInfos, stream)
+
+	log.Info("create message stream",
+		zap.Int64("collection_id", collectionID),
+		zap.Strings("virtual_channels", channelInfos.vchans),
+		zap.Strings("physical_channels", channelInfos.pchans))
+
+	for _, pc := range channelInfos.pchans {
 		metrics.ProxyMsgStreamObjectsForPChan.WithLabelValues(strconv.FormatInt(Params.ProxyCfg.GetNodeID(), 10), pc).Inc()
 	}
+
 	return nil
 }
 
-func (mgr *singleTypeChannelsMgr) getStream(collectionID UniqueID) (msgstream.MsgStream, error) {
-	mgr.streamMtx.RLock()
-	defer mgr.streamMtx.RUnlock()
+func (mgr *singleTypeChannelsMgr) lockGetStream(collectionID UniqueID) (msgstream.MsgStream, error) {
+	mgr.mu.RLock()
+	defer mgr.mu.RUnlock()
+	streamInfos, ok := mgr.infos[collectionID]
+	if ok {
+		return streamInfos.stream, nil
+	}
+	return nil, fmt.Errorf("collection not found: %d", collectionID)
+}
 
-	vid, err := mgr.getLatestVID(collectionID)
-	if err != nil {
-		return nil, err
+// getStream get message stream of specified collection.
+// If stream don't exists, call createMsgStream to create for it.
+func (mgr *singleTypeChannelsMgr) getStream(collectionID UniqueID) (msgstream.MsgStream, error) {
+	if stream, err := mgr.lockGetStream(collectionID); err == nil {
+		return stream, nil
 	}
 
-	stream, ok := mgr.id2Stream[vid]
-	if !ok {
-		return nil, fmt.Errorf("no dml stream for collection %v", collectionID)
+	if err := mgr.createMsgStream(collectionID); err != nil {
+		return nil, err
 	}
 
-	return stream, nil
+	return mgr.lockGetStream(collectionID)
 }
 
+// removeStream remove the corresponding stream of the specified collection. Idempotent.
+// If stream already exists, remove it, otherwise do nothing.
 func (mgr *singleTypeChannelsMgr) removeStream(collectionID UniqueID) error {
-	channels, err := mgr.getChannels(collectionID)
-	if err != nil {
-		return err
-	}
-	ids, err2 := mgr.getAllVIDs(collectionID)
-	if err2 != nil {
-		return err2
-	}
-
-	mgr.deleteVChansByVIDs(ids)
-	mgr.deleteStreamByVIDs(ids)
-	for _, pc := range channels {
-		metrics.ProxyMsgStreamObjectsForPChan.WithLabelValues(strconv.FormatInt(Params.ProxyCfg.GetNodeID(), 10), pc).Dec()
-	}
+	mgr.mu.Lock()
+	defer mgr.mu.Unlock()
+	// no need to check if collection exists.
+	delete(mgr.infos, collectionID)
 	return nil
 }
 
+// removeAllStream remove all message stream.
 func (mgr *singleTypeChannelsMgr) removeAllStream() error {
-	mgr.deleteAllChannels()
-	mgr.deleteAllStream()
-	mgr.deleteAllVChans()
-	mgr.deleteAllCollection()
-
+	mgr.mu.Lock()
+	defer mgr.mu.Unlock()
+	mgr.infos = make(map[UniqueID]streamInfos)
```

Comment:
call Close of  All MsgStream Object to release resource.

Reply:
We'll close MsgStream in it's Finalizer.